### DISCUSSION TITLE
Add A Custom VertexElement

### DISCUSSION THREAD

**Discussion by Bodyclock-Games** [stride3d/stride#1874] [Q&A] [answered]
Hi All, a newcomer to Stride from Unity and I'm looking to port some of my C# modules to Stride. One of these is a terrain system that uses custom VertexAttributes to store data such as slope and aspect in a Vector4. Is it possible to add a custom VertexElement type to the VertexElement struct or is that structure constrained to the requirements of DX11. This is lower level code than I'm used to and I don't want to mess this up. 
Or can I just declare an additional  TextureCoordinate<Vector4> in a custom VertexPositionNormalColor type struct.
Or alternatively is there a better way to do this?
Cheers
Chris

**Comment by NicusorN5** [comment]
Hi, a starting point is ths: https://doc.stride3d.net/4.1/en/manual/graphics/low-level-api/draw-vertices.html

This article describes creating a vertex buffer, a vertex declaration and finally drawing the vertex buffer.

**Comment by ly3027929699** [comment, marked as answer]
if you want to define an ext texturecoord, write a custom shader
![%2C4BTQKAI {GL38NW~5V60](https://github.com/stride3d/stride/assets/42492716/1c735f33-fbbe-4793-98bd-6faa02f3ed80)
![6P(C)VDTK~WP(FH~5C~2G~4](https://github.com/stride3d/stride/assets/42492716/b579c0fc-4f46-4667-a98a-5bda9749ea15)

**Comment by Bodyclock-Games** [comment]
Many `thanks` for the answers. This is what I have so far. Does this look reasonable?

`using System;
using System.Runtime.InteropServices;
using Stride.Graphics;
using Stride.Core.Mathematics;

namespace Bodyclock.Stride.Terrain
{
    [StructLayout(LayoutKind.Sequential, Pack = 4)]
    public struct TerrainVertex : IEquatable<TerrainVertex>, IVertex
    {
        public Vector3 Position;
        public Vector3 Normal;
        public Vector4 Attributes;
        public Color Color;

        public static readonly int Size = Marshal.SizeOf(typeof(TerrainVertex));

        public static readonly VertexDeclaration Layout = new(
        VertexElement.Position<Vector3>(),
        VertexElement.Normal<Vector3>(),
        new VertexElement("ATTRIBUTES", PixelFormat.R32G32B32A32_Float),
        VertexElement.Color<Color>());

        public TerrainVertex(Vector3 position, Vector3 normal, Vector4 attributes, Color color)
        {
            this = default;
            Position = position;
            Normal = normal;
            Attributes = attributes;
            Color = color;
        }

        public VertexDeclaration GetLayout() => Layout;
        public void FlipWinding() { }
        public bool Equals(TerrainVertex other)
        {
            if (Position.Equals(other.Position) && Normal.Equals(other.Normal) && Attributes.Equals(other.Attributes))
                return Color.Equals(other.Color);
            return false;
        }

        public override bool Equals(object obj)
        {
            if (obj == null) return false;
            if (obj is TerrainVertex vertex) return Equals(vertex);
            return false;
        }

        public override int GetHashCode()
        {
            unchecked
            {
                int hashCode = Position.GetHashCode();
                hashCode = (hashCode * 397) ^ Normal.GetHashCode();
                hashCode = (hashCode * 397) ^ Attributes.GetHashCode();
                hashCode = (hashCode * 397) ^ Color.GetHashCode();
                return hashCode;
            }
        }
        public static bool operator ==(TerrainVertex left, TerrainVertex right)
        {
            return left.Equals(right);
        }
        public static bool operator !=(TerrainVertex left, TerrainVertex right)
        {
            return !left.Equals(right);
        }
        public override string ToString()
        {
            return $"Position: {Position}, Normal: {Normal}, Attributes: {Attributes}, Color: {Color}";
        }
    }
}`

**Comment by ly3027929699** [comment]
using the TerrainVertex.Layout when creating a vetexbuffer, and it' will be done. link: https://doc.stride3d.net/4.1/en/manual/graphics/low-level-api/draw-vertices.html. Also, you can check the UnitTest of stride's source. It hsa some samples of drawing

**Reply by ly3027929699** [reply]
![U$1RX(E_LQOPZX7}@{{9Q55](https://github.com/stride3d/stride/assets/42492716/dcaac53c-9287-4b4c-b0e8-c6d0e44f45ac)

**Comment by Ethereal77** [comment]
This would be a nice thing to have in the docs.
